### ISSUE TITLE
ERROR:Daemon already running: pid 1603

### ISSUE THREAD

**Issue by magnetophon** [mrichar1/clipster#77]
Sometimes after a reboot, the daemon systemd service I set up doesn't start, and when I try manually, I get the above error.
The pid is different every boot, and always actually taken by a systemd related thing.

I've never noticed any other services not comig up because of this, but I'm not running many either.

I don't know if this is a clipster issue,or something with systemd or my distro.

Any ideas how to troubleshoot?

**Comment by mrichar1** [comment]
This sounds like a systemd rather than a clipster issue - can you share your systemd script? I have a vague memory of there being a mechanism where systemd can generate pids for processes which don't have process management builtin - perhaps this is what's causing the unusual behaviour here?

**Comment by magnetophon** [comment]
[Here's the code](https://github.com/magnetophon/nixosConfig/blob/master/common.nix#L519-L530)

**Comment by mrichar1** [comment]
Does the conflicting PID point to a running clipster process (i.e it's being started twice) or is it an 'old' PID from a previous instance?

If it's the latter, it might help to add the `PIDFile` option to the systemd service to explicitly name the pid file clipster will be using, so systemd can tidy it up if it's left lying around:

```
       PIDFile=
           Takes an absolute file name pointing to the PID file of this daemon. Use of this
           option is recommended for services where Type= is set to forking. systemd will
           read the PID of the main process of the daemon after start-up of the service.
           systemd will not write to the file configured here, although it will remove the
           file after the service has shut down if it still exists.
```

**Comment by magnetophon** [comment]
I [tried to set a PIDFile](https://github.com/magnetophon/nixosConfig/blob/master/common.nix#L534) but I'm not sure if I was succesfull.
Is there a way to check?

**Comment by magnetophon** [comment]
I just ran into this issue again, while still using the above attempt at a solution.

**Comment by mrichar1** [comment]
Can you confirm whether or not the pid specified in the `daemon already running` message corresponds to a running process, or is there a pid file but no associated process?

**Comment by magnetophon** [comment]
Yes, the PID always corresponds to a running process, started by systemd.

**Comment by mrichar1** [comment]
If there is a running process, started by systemd, then for some reason it sounds like systemd is trying to start the daemon twice.Can you provide the resulting systemd service that your config generates (`systemctl show clipster.service`)

At this point it sounds like it's systemd rather than clipster specific - so I'm not sure I can offer much more advice - not being a systemd expert. The only other thing to do (other than sharing the service file)  is check that, since it it appears your config is generating a user systemd service, whether this is trying to start clipster for multiple users or multiple logins without modifying the pidfile/history etc paths?

**Comment by magnetophon** [comment]
I don't have any other users on this system, other than root.
The PID in the error isn't used bu clipster, but always by another process.
 
```
systemctl show clipster.service                                                                                                                                                          Restart=no
NotifyAccess=none
RestartUSec=100ms
TimeoutStartUSec=1min 30s
TimeoutStopUSec=1min 30s
RuntimeMaxUSec=infinity
WatchdogUSec=0
WatchdogTimestampMonotonic=0
PermissionsStartOnly=no
RootDirectoryStartOnly=no
RemainAfterExit=no
GuessMainPID=yes
MainPID=0
ControlPID=0
FileDescriptorStoreMax=0
NFileDescriptorStore=0
StatusErrno=0
Result=success
UID=[not set]
GID=[not set]
NRestarts=0
ExecMainStartTimestampMonotonic=0
ExecMainExitTimestampMonotonic=0
ExecMainPID=0
ExecMainCode=0
ExecMainStatus=0
MemoryCurrent=[not set]
CPUUsageNSec=[not set]
TasksCurrent=[not set]
IPIngressBytes=18446744073709551615
IPIngressPackets=18446744073709551615
IPEgressBytes=18446744073709551615
IPEgressPackets=18446744073709551615
Delegate=no
CPUAccounting=no
CPUWeight=[not set]
StartupCPUWeight=[not set]
CPUShares=[not set]
StartupCPUShares=[not set]
CPUQuotaPerSecUSec=infinity
IOAccounting=no
IOWeight=[not set]
StartupIOWeight=[not set]
BlockIOAccounting=no
BlockIOWeight=[not set]
StartupBlockIOWeight=[not set]
MemoryAccounting=yes
MemoryLow=0
MemoryHigh=infinity
MemoryMax=infinity
MemorySwapMax=infinity
MemoryLimit=infinity
DevicePolicy=auto
TasksAccounting=yes
TasksMax=4915
IPAccounting=no
UMask=0022
LimitCPU=infinity
LimitCPUSoft=infinity
LimitFSIZE=infinity
LimitFSIZESoft=infinity
LimitDATA=infinity
LimitDATASoft=infinity
LimitSTACK=infinity
LimitSTACKSoft=8388608
LimitCORE=infinity
LimitCORESoft=infinity
LimitRSS=infinity
LimitRSSSoft=infinity
LimitNOFILE=1048576
LimitNOFILESoft=1048576
LimitAS=infinity
LimitASSoft=infinity
LimitNPROC=31173
LimitNPROCSoft=31173
LimitMEMLOCK=16777216
LimitMEMLOCKSoft=16777216
LimitLOCKS=infinity
LimitLOCKSSoft=infinity
LimitSIGPENDING=31173
LimitSIGPENDINGSoft=31173
LimitMSGQUEUE=819200
LimitMSGQUEUESoft=819200
LimitNICE=0
LimitNICESoft=0
LimitRTPRIO=0
LimitRTPRIOSoft=0
LimitRTTIME=infinity
LimitRTTIMESoft=infinity
OOMScoreAdjust=0
Nice=0
IOSchedulingClass=0
IOSchedulingPriority=0
CPUSchedulingPolicy=0
CPUSchedulingPriority=0
TimerSlackNSec=50000
CPUSchedulingResetOnFork=no
NonBlocking=no
StandardInput=null
StandardInputData=
StandardOutput=inherit
StandardError=inherit
TTYReset=no
TTYVHangup=no
TTYVTDisallocate=no
SyslogPriority=30
SyslogLevelPrefix=yes
SyslogLevel=6
SyslogFacility=3
LogLevelMax=-1
SecureBits=0
CapabilityBoundingSet=cap_chown cap_dac_override cap_dac_read_search cap_fowner cap_fsetid cap_kill cap_setgid cap_setuid cap_setpcap cap_linux_immutable cap_net_bind_service cap_net_broadcast cap_n>
AmbientCapabilities=
DynamicUser=no
RemoveIPC=no
MountFlags=
PrivateTmp=no
PrivateDevices=no
ProtectKernelTunables=no
ProtectKernelModules=no
ProtectControlGroups=no
PrivateNetwork=no
PrivateUsers=no
ProtectHome=no
ProtectSystem=no
SameProcessGroup=no
UtmpMode=init
IgnoreSIGPIPE=yes
NoNewPrivileges=no
SystemCallErrorNumber=0
LockPersonality=no
RuntimeDirectoryPreserve=no
RuntimeDirectoryMode=0755
StateDirectoryMode=0755
CacheDirectoryMode=0755
LogsDirectoryMode=0755
ConfigurationDirectoryMode=0755
MemoryDenyWriteExecute=no
RestrictRealtime=no
RestrictNamespaces=no
MountAPIVFS=no
KeyringMode=private
KillMode=control-group
KillSignal=15
SendSIGKILL=yes
SendSIGHUP=no
Id=clipster.service
Names=clipster.service
Description=clipster.service
LoadState=not-found
ActiveState=inactive
SubState=dead
StateChangeTimestampMonotonic=0
InactiveExitTimestampMonotonic=0
ActiveEnterTimestampMonotonic=0
ActiveExitTimestampMonotonic=0
InactiveEnterTimestampMonotonic=0
CanStart=no
CanStop=yes
CanReload=no
CanIsolate=no
StopWhenUnneeded=no
RefuseManualStart=no
RefuseManualStop=no
AllowIsolate=no
DefaultDependencies=yes
OnFailureJobMode=replace
IgnoreOnIsolate=no
NeedDaemonReload=no
JobTimeoutUSec=infinity
JobRunningTimeoutUSec=infinity
JobTimeoutAction=none
ConditionResult=no
AssertResult=no
ConditionTimestampMonotonic=0
AssertTimestampMonotonic=0
LoadError=org.freedesktop.DBus.Error.FileNotFound "No such file or directory"
Transient=no
Perpetual=no
StartLimitIntervalUSec=10s
StartLimitBurst=5
StartLimitAction=none
FailureAction=none
SuccessAction=none
CollectMode=inactive
```

**Comment by mrichar1** [comment]
OK - I went away and thought about this, and re-read the issue, and realised what I think the problem is.

Clipster stores it's pid file in `data_dir/clipster.pid` by default - this is usually somewhere like `/home/user/.local/share/clipster/clipster.pid`

Now, if clipster fails to shutdown successfully, then the pid file remains intact. For example, a hard shutdown. On reboot, this file still exists, so clipster looks to see if there is a process with that pid running. Often there will be, so it refuses to start.

There are probably 2 ways to try to fix this.
The first is to save the pid file somewhere that gets cleared out on reboot, the second is to have clipster check that, if the pid exists, it belongs to a clipster process, rather than just stopping if any process has that pid.

The first fix is an easy one to test. Edit your config and add:

```
pid_file = /run/user/1000/clipster.pid
```

(You'll need to replace `1000` with your uid: `echo $UID` should give you this).

Then kill clipster uncleanly - either `kill -9` or pull the power on your computer. Now the pid file should have gone away and clipster can't complain about existing processes.

If this works I'll see about making that path the default for clipster's pid file anyway.

**Comment by magnetophon** [comment]
Thanks!
I edited my configfile, and did ``kill -9 $PID-of-clipster`` and systemd restarted clipster.
I'm not sure if that means the test was succesfull.

**Comment by mrichar1** [comment]
I've now made the default `pid_file` location `/run/user/$UID/clipster.pid`, since this should be properly tidied up on reboot (this can be overridden in the config if the old behaviour is needed.

Hopefully this will ensure that there can't be a stale pid_file on the system after a reboot.

Fixed in 2.0.0 release.
